File: haskell/ss5/sortsum.hs
import Data.List

sortsum :: (Ord a, Num a) => [a] -> [a] -> [a]
sortsum xs ys = sort [x + y | x <- xs, y <- ys]

-- x - y = x + negate y

type Label a = (a,(Int,Int))

subs :: (Num a) => [a] -> [a] -> [Label a]
subs xs ys = [(x - y, (i,j)) | (x,i) <- zip xs [1..],(y,j) <- zip ys [1..]]

sortsums xs ys = map fst (sortsubs xs (map negate ys))
sortsubs xs ys = sort (subs xs ys)

table :: (Ord a,Num a) => [a] -> [a] -> [(Int,Int,Int)]
table xs ys = map snd (map (tag 1) xxs `merge` map (tag 2) yys)
              where xxs = sortsubs xs xs
                    yys = sortsubs ys ys
tag i (x,(j,k)) = (x,(i,j,k))

merge :: (Ord a) => [a] -> [a] -> [a]
merge [] ys = ys
merge xs [] = xs
merge (x:xs) (y:ys) | x < y     = x : merge xs (y:ys)
                    | otherwise = y : merge (x:xs) ys

File: src/main/scala/examples/ss5/Sortsum.scala
package examples.ss5

object Sortsum {

  type Label = (Int,(Int,Int))
  type TagedLabel = (Int,(Int,Int,Int))
    
  def sortsum(xs: List[Int], ys: List[Int]): List[Int] = 
    (for (x <- xs; y <- ys) yield x + y).sorted
  
  def subs(xs: List[Int], ys: List[Int]): List[Label] = {
    for {
      (i, x) <- Stream.from(1).zip(xs).toList
      (j, y) <- Stream.from(1).zip(ys).toList
    } yield (x - y, (i,j))
  }
  
  def sortsums(xs: List[Int], ys: List[Int]) = sortsubs(xs, ys.map(_ * -1)).map(_._1)
  
  def sortsubs(xs: List[Int], ys: List[Int]):List[Label] = subs(xs,ys).sorted
  
  def merge[A: Ordering](xs: List[A], ys: List[A]): List[A] = {
    val x :: xxs = xs
    val y :: yys = ys
    if (implicitly[Ordering[A]].lt(x,y)) x :: merge(xxs, ys)
    else y :: merge(xs, yys)
  } 
  
  def tag(i:Int, x: Label): TagedLabel = (x._1, (i,x._2._1, x._2._1))
      
  def table(xs: List[Int], ys: List[Int]):List[(Int,Int,Int)] = {
    val xxs:List[TagedLabel] = sortsubs(xs,xs).map(tag(1,_))
    val yys:List[TagedLabel] = sortsubs(ys,ys).map(tag(2,_))
    merge(xxs, yys).map(_._2)
  } 
}